doc/module.1.in
.\" .TH MODULE 1 "10 October 2005
.\"  minor editorial edits grenker 20090711
.TH MODULE 1 "2020-04-07" "Modules version @VERSION@" "Modules package"
.nh
.SH NAME
module \- command interface to the Modules package
.SH SYNOPSIS
.B module
[
.I switches
]
[
.I sub-command
]
.I [ sub-command-args ]
.SH DESCRIPTION
.LP
.B module
is a user interface to the Modules package.
The Modules package provides for the dynamic modification of the user's
environment via
.I modulefiles.
.LP
Each
.I modulefile
contains the information needed to configure the shell for an application.
Once the Modules package is initialized, the environment can be modified
on a per-module basis using the 
.B module
command which interprets
.I modulefiles.
Typically
.I modulefiles
instruct the
.B module
command to alter or set shell environment variables such as PATH, MANPATH,
etc.
.I modulefiles
may be shared by many users on a system and users may have their own
collection to supplement or replace the shared
.I modulefiles.
.LP
The
.I modulefiles
are added to and removed from the current
environment by the user.
The environment changes contained in a
.I modulefile
can be summarized through the
.B module
command as well.
If no arguments are given, a summary of the
.B module
usage and
.I sub-commands
are shown.
.LP
The action for the
.B module
command to take is described by the
.I sub-command
and its associated arguments.
.SS Package Initialization
.LP
The Modules package and the
.B module
command are initialized when a shell-specific initialization script is
sourced into the shell.
The script creates the
.B module
command, either as an alias or shell function,
creates Modules environment variables,
and if enabled to do so, a snapshot of the environment is saved as either
(if BEGINENV=1)
.I $HOME/.modulesbeginenv
or (if BEGINENV=99) whatever
.I $MODULESBEGINENV
points to.

The
.B module
alias or function executes the
.B modulecmd
program and has the shell evaluate
the command's output.  The first argument to
.B modulecmd
specifies the type of shell.
.LP
The initialization scripts are kept in
.I $MODULESHOME/init/<shell>
where
.I <shell>
is the name of the sourcing shell.
The sh, csh, tcsh, bash, ksh, and zsh shells are supported by
.B modulecmd.
In addition, python, perl, and cmake "shells" are supported, which writes
the environment changes to stdout as python, perl, or cmake code.

The perl
.I module
command is set up with:

  use lib $ENV{'MODULESHOME'}."/init";
  use perl;

And the python
.I module
command is defined with:

  import os, sys;
  sys.path.insert(0, '@INITPATH@')

  from python import module;

.SS Modulecmd startup
.LP
Upon invocation \fImodulecmd\fP sources rc files which contain global, user
and modulefile specific setups. These files are interpreted as modulefiles.
See
.BR modulefile (4)
for detailed information.
.LP
Upon invocation of modulecmd module RC files are sourced in the following
order:
.nf

	Global RC file as specified by ${\s-1MODULERCFILE\s0} or ${\s-1MODULESHOME\s0}/etc/rc

	User specific module RC file ${\s-1HOME\s0}/.modulerc

	All .modulerc and .version files found during modulefile seeking.

.fi
.SS Command line switches
.LP
The module command accepts command line switches as its first parameter. These
may be used to control output format of all information displayed and the 
\fImodule\fP behavior in case of locating and interpreting module files.
.LP
All switches may be entered either in short or long notation. The following
switches are accepted:
.PP
.TP
.I --help, -H
Give some helpful usage information, and terminates the command.
.RS
.RE
.TP
.I --version, -V
Lists the current version of the module command, and some configured
option values.  The command then terminates without further processing.
.RS
.RE
.TP
.I --force, -f
Force active dependency resolution. This will result in modules found on a
\fIprereq\fP command inside a module file being load automatically.
Unloading module files using this switch will result in all required modules
which have been loaded automatically using the \fI-f\fP switch being unload.
This switch is experimental at the moment.
.RS
.RE
.TP
.I --terse, -t
Display \fIavail\fP and \fIlist\fP output in short format.
.RS
.RE
.TP
.I --long, -l
Display \fIavail\fP and \fIlist\fP output in long format.
.RS
.RE
.TP
.I --human, -h
Display short output of the \fIavail\fP and \fIlist\fP commands in human
readable format. 
.RS
.RE
.TP
.I --verbose, -v
Enable verbose messages during module command execution.
.RS
.RE
.TP
.I --silent, -s
Disable verbose messages. Redirect \fIstderr\fP to \fI/dev/null\fP if 
\fIstderr\fP is found not to be a tty. This is a useful option for module
commands being written into \fI.cshrc\fP, \fI.login\fP or \fI.profile\fP
files, because some remote shells (as \fIrsh\fP(1)) and remote execution
commands (like \fIrdist\fP) get confused if there is output on stderr.
.RS
.RE
.TP
.I --create, -c
Create caches for \fImodule avail\fP and \fImodule apropos\fP. You must 
be granted write access to the ${\s-1MODULEHOME\s0}/modulefiles/ directory
if you try to invoke \fImodule\fP with the \fI-c\fP option.
.RS
.RE
.TP
.I --icase, -i
Case insensitive module parameter evaluation. Currently only implemented for
the \fImodule apropos\fP command.
.RS
.RE
.TP
.I --userlvl <lvl>, -u <lvl>
Set the user level to the specified value. The argument of this option may
be one of:
.LP
.I novice, nov
Novice
.LP
.I expert, exp
Experienced module user
.LP
.I advanced, adv
Advanced module user
.SS Module Sub-Commands
.RS
.TP 15
.B help\ \  [modulefile...]
Print the usage of each sub-command.
If an argument is given, print the Module-specific help information for the
\fImodulefile\fP(s).
.TP 15
.B add\ \ \  modulefile...
.PD 0
.TP 15
.B load\ \  modulefile...
.PD
Load
.IR modulefile (s)
into the shell environment.
.TP 15
.B rm\ \ \ \  modulefile...
.PD 0
.TP 15
.B unload modulefile...
.PD
Remove
.IR modulefile (s)
from the shell environment.
.TP 15
.B swap\ \  [modulefile1] modulefile2
.PD 0
.TP 15
.B switch [modulefile1] modulefile2
.PD
Switch loaded \fImodulefile1\fP with \fImodulefile2\fP.
If \fI modulefile1\fP is not specified, then it is assumed to be the currently loaded module
with the same root name as \fImodulefile2\fP.
.TP 15
.B show\ \ \  modulefile...
.PD 0
.TP 15
.B display modulefile...
.PD
Display information about one or more
.IR modulefile s.
The display sub-command will list the full path of the \fImodulefile\fP(s)
and all (or most) of the environment changes the modulefile(s)
will make if loaded.  (It will not display any
environment changes found within conditional statements.)
.TP 15
.B list
List loaded modules.
.TP 15
.B avail [path...]
List all available modulefiles in the current \s-1MODULEPATH\s0, where
the sorting order is given by the \s-1LC_COLLATE\s0 locale environment
variable.

All directories in the \s-1MODULEPATH\s0 are recursively searched for
files containing the \fImodulefile\fP magic cookie.

If an argument is given, then each directory in the \s-1MODULEPATH\s0 is
searched for modulefiles whose pathname match the argument.

Multiple versions of an application can be supported by creating a
subdirectory for the application containing modulefiles for each version.
.PD
.TP 15
.B use [-a|--append] directory...
.PD
Prepend one or more directories to the \s-1MODULEPATH\s0 environment variable.
The --append flag will append the directory to \s-1MODULEPATH\s0.
.TP 15
.B unuse directory...
Remove one or more directories from the \s-1MODULEPATH\s0 environment variable.
.TP 15
.B update
Attempt to reload all loaded modulefiles.
The environment will be reconfigured to match the environment saved in
.I ${\s-1HOME\s0}/.modulesbeginenv
(if BEGINENV=1)
or the file pointed at by
.I $MODULESBEGINEV
(if BEGINENV=99)
and the modulefiles will be reloaded.
This is only valid if modules was configured with --enable-beginenv
(which defines BEGINENV), otherwise this will cause a warning.
.B update
will only change the environment variables that the
.IR modulefile s
set.
.TP 15
.B clear
Force the Modules package to believe that no modules are currently loaded.
.TP 15
.B purge
Unload all loaded modulefiles.
.TP 15
.B refresh
Force a refresh of all non-persistent components of currently loaded modules.
This should be used on derived shells where aliases need to be reinitialized
but the environment variables have already been set by the currently loaded
modules.
.TP 15
.B whatis [modulefile...]
Display the information set up by the \fImodule-whatis\fP commands
inside the specified modulefile(s). If no \fImodulefile\fP is specified,
all 'whatis' lines will be shown.
.TP 15
.B apropos string
.PD 0
.TP 15
.B keyword string
Seeks through the 'whatis' informations of all modulefiles for the specified string.
All \fImodule-whatis\fP informations matching the \fIstring\fP will be displayed.
.PD
.TP 15
.B initadd modulefile...
Add
.IR modulefile (s)
to the shell's initialization file in the user's home directory.
The startup files checked (in order) are:
.br
csh - .modules, .cshrc(.ext), .csh_variables, and .login(.ext)
.br
tcsh - .modules, .tcshrc, .cshrc(.ext), .csh_variables, and .login(.ext)
.br
sh and ksh - .modules, .profile(.ext), and .kshenv(.ext)
.br
bash - .modules, .bash_profile, .bash_login, .profile(.ext), and .bashrc(.ext)
.br
zsh - .modules, .zcshrc(.ext), .zshenv(.ext), and .zlogin(.ext)

If a 'module load' line is found in any of these files, the
\fImodulefile\fP(s) is(are) appended to any existing list of
modulefiles.
The 'module load' line must be located in at least one of the
files listed above for any of the 'init' sub-commands to work properly.
If the 'module load' line is found in multiple shell initialization files,
all of the lines are changed.
.TP 15
.B initprepend modulefile [modulefile...]
Does the same as
.B initadd
but prepends the given modules to the beginning of the list.
.TP 15
.B initrm modulefile...
Remove \fImodulefile\fP(s) from the shell's initialization files.
.TP 15
.B initswitch modulefile1 modulefile2
Switch
.I modulefile1
with
.I modulefile2
in the shell's initialization files.
.TP 15
.B initlist
List all of the modulefiles loaded from the shell's initialization file.
.TP 15
.B initclear
Clear all of the modulefiles from the shell's initialization files.
.RE
.SS Modulefiles
.LP
.IR modulefile s
are written in the Tool Command Language (Tcl) and are interpreted by
.B modulecmd.
.IR modulefile s
can use conditional statements.
Thus the effect a 
.I modulefile
will have on the environment may change depending upon the current state of
the environment.
.LP
Environment variables are unset when unloading a
.I modulefile.
Thus, it is possible to
.B load
a 
.I modulefile
and then
.B unload
it without having the environment variables return to their
prior state.
.LP
.SH ENVIRONMENT
.TP
.SB \s-1MODULESHOME\s0
The location of the master Modules package file directory containing
.B module
command initialization scripts, the executable program
.B modulecmd,
and a directory containing a collection of master
.I modulefiles.
.TP
.SB \s-1MODULEPATH\s0
The path that the
.B module
command searches when looking for
.I modulefiles.
Typically, it is set to a default value by the bootstrap procedure.
\s-1MODULEPATH\s0 can be set using 'module use' or by the module
initialization script to search group or personal
.I modulefile
directories before or after the master
.I modulefile
directory.
.TP
.SB \s-1LOADEDMODULES\s0
A colon separated list of all loaded
.I modulefiles.
.TP
.SB \s-1_LMFILES_\s0
A colon separated list of the full pathname for all loaded
.I modulefiles.
.TP
.SB \s-1MODULESBEGINENV\s0
If modules has been configured (BEGINENV=99) to test for this environment
variable, then if it exists, it is the name of the file to store the
the initial shell environment.  This environment variable will have
embedded environment variables unrolled to one level.  The contents of
this variable is only used the first time
.I modules
is invoked.
.TP
.SB \s-1_MODULESBEGINENV_\s0
The filename of the file containing the initialization environment snapshot.

.SH FILES
.TP
.B @prefix@
The \s-1MODULESHOME\s0 directory.
.TP
.B ${\s-1MODULESHOME\s0}/etc/rc
The system-wide
.I modules rc file.
The location of this file can be changed using
the \s-1MODULERCFILE\s0 environment variable as described above.
.TP
.B ${\s-1HOME\s0}/.modulerc
The user specific 
.I modules rc file.
.TP
.B ${\s-1MODULESHOME\s0}/modulefiles
The directory for system-wide
.I modulefiles.
The location of the directory can be changed using
the \s-1MODULEPATH\s0 environment variable as described above.
.TP
.B ${\s-1MODULESHOME\s0}/bin/modulecmd
The 
.I modulefile
interpreter that gets executed upon each invocation of
.B module.
.TP
.B ${\s-1MODULESHOME\s0}/init/\fI<shell>\fP
The Modules package initialization file sourced into the user's
environment.
.TP
.B ${\s-1MODULEPATH\s0}/.moduleavailcache
File containing the cached list of all
.I modulefiles
for each directory in the \s-1MODULEPATH\s0
(only when the avail cache is enabled via the configure option --enable-cache
which sets CACHE_AVAIL).
.TP
.B ${\s-1MODULEPATH\s0}/.moduleavailcachedir
File containing the names and modification times for all sub-directories with
an avail cache (see above).
.TP
.B ${\s-1HOME\s0}/.modulesbeginenv
A snapshot of the user's environment taken at Module initialization.
This information is used by the
.B module update
sub-command (if BEGINENV=1), else
.TP
.B $MODULESBEGINENV
If this defines a valid filename, it serves the same purpose as above
(if BEGINENV=99).
.SH VERSION
@VERSION@
.SH SEE ALSO
.BR modulefile (4)


NEWS
* Note: Implied Consent -
  If you send in a code patch or fix of one type or another you consent
  to have your name and email address listed here unless you explicitly
  note otherwise.  Whether your name & email is actually listed depends
  on how extensive the change.  (If it's a minor code, document change,
  or spelling correction your name will probably not be listed.)

****************************   Release 3.2 ******************************
Modules 3.2.13
Apr  7, 2020				Xavier Delaruelle (xavier.delaruelle@cea.fr)
    * Fix buffer overflow when working with large MODULEPATH (thanks to
      Dominique Martinet for his help) (fix issue#311)
    * Install: dynamically set gitlog2changelog.py script shebang
    * Fix unset-alias logic (Martin Schuster) (fix issue#321)

Modules 3.2.12
Mar 23, 2019				Xavier Delaruelle (xavier.delaruelle@cea.fr)
    * Fix some C-string issues (Tim Haines) (fix issue#221)
    * Fix configure.ac not to search for a TclX installation if
      --without-tclx is set (fix issue#217)

Modules 3.2.11
Oct 16, 2017				Xavier Delaruelle (xavier.delaruelle@cea.fr)
	* Fixed configure.ac bug with AM_INIT_AUTOMAKE (Alexey Brodkin)
	* Fix compilation with -Werror=implicit-function-declaration
	  (Orion Poplawski)
	* Fix default version set in nested directory (Jan Synacek)
	* Make Python init script Python 3 compatible (Slavek Kabrda)
	* Fix unload from loaded modulefile (Filip Krska)
	* Add support for Tcl 8.6 (Orion Poplawski)
	* Fix build with -Werror=format-security (Orion Poplawski)
	* Fix minor compiler warnings (Alastair McKinstry)
	* Add completion to avail command on bash (Orion Poplawski)
	* Fix "module clear" command (Orion Poplawski)
	* Comment out stray "module use" in example modules file when not
	  using versioning (Orion Poplawski)
	* Fix modulecmd path in init files (Orion Poplawski)

Modules 3.2.10
Dec 21, 2012				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed the module switch with custom delimiters (Tyson Whitehead)
	* If a bash shell check if interactive or not for alias/functions
	* Fix the flags with regards to recursive load/unload
	* Fix perl.pm init file typo (bug 3497092)
	* Moved from cvs to git (and now: modulefile/module-git)
	* Fix the module swap with the --delim option (Tyson Whitehead)
	* Increased the "update" buffer (Leon Kos)
	* Use the locale for sorting modulefiles by the avail command
	* Fixed the segfault problem with Tcl_RegExpCompile() (Kenneth Hoste,
	  Tyson Whitehead)
	* Updated the FSF address in LICENSE.GPL (Jan Synacek)
	* Forked the Tcl-only version to be maintained by Kent Mein.
	* Added the "module-info type" query to return 'C' to indicate that
	  the modulecmd executable is the "C" version.
	  (The Tcl-only version will return 'Tcl'.)
	* Search /usr/lib64 for tclConfig.sh also (Jan Synacek)
	* Tested loading 20 other modulefiles from a single one
	* Have ChangeLog generated by a script due to Marcus D. Hanwell

Modules 3.2.9
Nov 24, 2011				R.K. Owen (rk@owen.sj.ca.us)
	* Fix the coexistence of load/remove flags (Martin Siegert)
	* Fixed an overlapping string copy (Wilson Snyder & Orion Poplawski)
	* Isolated calls to the new interpreter and get & set env.vars.
	* Crucial bug fix of same by (Orion Poplawski).
	* Optionally use Tcl memory check routines for malloc/realloc.
	* Fixed the "module purge" memory corruption (Poor Yorick)
	* Added support for Ruby (Tammo Tjarks)

Modules 3.2.8
Oct  1, 2010				R.K. Owen (rk@owen.sj.ca.us)
	* Handle the .modulerc/.version in deep modulefile dirs,
	  Patches thanks to Lindsay Todd, and others.
	* Note that .version takes precedence over .modulerc files.
	* Renamed init/python to init/python.py and updated with
	  anonymous patch to eliminate 'sed' references
	* Renamed init/perl to init/perl.pm
	* Added brief section to module.1 on invoking modules for perl & python
	* Added Nathaniel Waisbrot's patch for CMake support with tests.
	* Fixed the MANPATH in modulefiles/modules to use configure generated
	  values.
	* Added the configure --with-man-path so a system specific MANPATH
	  could be used.  Some "man"s radically change their search path
	  when a MANPATH exists and when not.
	  The default is "/usr/man:/usr/share/man" (when both exists).
	* Purging a purged module list does not result in an error.
	* Fixed the configure script for searching for tclConfig.sh
	  and handle --without-x in the tests.

Modules 3.2.7
Jul 30, 2009				R.K. Owen (rk@owen.sj.ca.us)
	* Have init/bash export the module function (thanks to Roy Dragseth)
	* Fixed swap bug when prepending or appending multiple times
	  (fix thanks to Gavin Walker), added tests for same.
	* Have purge remove loaded modules in reverse order
	  (patch contributed by carriees)
	* Added a batch of patches from Gerrit Renker
	  - Goes through each of the two manpages, bringing them up to date.
	  - Fix an error in configure.ac caused by misspelling
	  - Fix typos in help messages and source-code comments
	  - Fix the setup of $MODULESPATH to not have a trailing ':'
	  - Fix segmentation violation occurring when the shelltype is absent
	  - Fix a condition where 'make uninstall' does not cleanse a directory
	  - Fix a build problem which set the executable bit on 'data' files
	  - Cleans up noise in '.modulespath' file
	  - Replaces external sed filter script with internal Makefile.am rule
	  - Adds missing 'zsh' case in etc/global
	  - simplify (t)csh initialisation
	  - Add bash auto-completion for the 'module' command
	  - Adds new 'chdir' modulefile command (sets cwd on module load)
	* versions a & b fixed some bugs

Modules 3.2.6
Oct 29, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed the locate_module.c:realloc() calls which didn't properly
	  take the element size into account (Fix thanks to Ron Isaacson).
Modules 3.2.5
Feb 14, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Wrapped all malloc() calls to make sure the size is never 0.
	  This fixes a problem spotted by Francesco Nigro under AIX,
	  and diagnosed by G Lehnert.

doc/source/module.pl
#!/usr/bin/perl

print "Run pod2html module.pl > ../html/module.html to generate html " .
	"documentation for the module command.\n";
print "Run pod2man module.pl > ../man/man1/module.1 to generate man " .
	"documentation for the module command.\n";

=head1 NAME

module - command interface to the Modules package

=head1 SYNOPSIS

module [switches] [sub-command [sub-command-args]]

=head1 DESCRIPTION

B<module> is a user interface to the Modules package.  The Modules package 
provides for the dynamic modification of the user's environment via I<modulefiles>.

Each I<modulefile> contains the information needed to configure the shell for
an application.  Once the Modules package is initialized, the environment can
be modified on a per-module basis using the B<module> command which interprets
I<modulefiles>.  Typically I<modulefiles> instruct the B<module> command to
alter or set shell environment variables such as PATH, MANPATH, etc.
I<Modulefiles> may be shared by many users on a system and users may have
their own collection to supplement or replace the shared I<modulefiles>.

The I<modulefiles> are added to and removed from the current environment by 
the user.  The environment changes contained in a I<modulefile> can be 
summarized through the B<module> command as well.
If no arguments are given, a summary of the B<module> usage and I<sub-commands>
are shown.

The action for the B<module> command to take is described by the I<sub-command>
and its associated arguments.

=head2 Package Initialization

The Modules package and the B<module> command are initialized when a 
shell-specific initialization script is sourced into the shell.
The script creates the B<module> command as either an alias or function,
creates Modules environment variables, and a snapshot of the environment
is saved in ${HOME}/.modulesbeginenv.  The B<module> alias or function
executes the B<modulecmd.tcl> program located in ${MODULESHOME}/ and has the
shell evaluate the command's output.  The first argument to B<modulecmd.tcl>
specifies the type of shell.

The initialization scripts are kept in ${MODULESHOME}/init/[shellname]
where [shellname] is the name of the sourcing shell.  For example, a C Shell
user sources the ${MODULESHOME}/init/csh script.  The sh, csh, tcsh, bash,
ksh, and zsh shells are supported by B<modulecmd.tcl>.  In addition, python and
perl "shells" are supported which writes the environment changes to stdout
as python or perl code.

=head2 Examples of initialization

In the following examples, replace ${MODULESHOME} with the actual directory
name.

C Shell initialization (and derivatives):
        source ${MODULESHOME}/init/csh
        module load modulefile modulefile ...

Bourne Shell (sh) (and derivatives):
        . ${MODULESHOME}/init/sh
        module load modulefile modulefile ...

Perl:
        require "${MODULESHOME}/init/perl";
        &module("load modulefile modulefile ...");

=head2 Modulecmd startup

Upon invocation B<modulecmd.tcl> sources rc files which contain global, user
and I<modulefile> specific setups. These files are interpreted as 
I<modulefiles>.  See I<modulefile> for detailed information.

Upon invocation of B<modulecmd.tcl> module RC files are sourced in the
following order:

        Global RC file as specified by ${MODULERCFILE} or 
		${MODULESHOME}/etc/rc

        User specific module RC file ${HOME}/.modulerc

        All .modulerc and .version files found during I<modulefile>
		seeking.

=head2 Command line switches

The B<module> command accepts command line switches as its first parameter.
These may be used to control output format of all information displayed and
the B<module> behavior in case of locating and interpreting I<modulefiles>.

All switches may be entered either in short or long notation. The following
switches are accepted:

=head3 --terse, -t

Display B<avail> and B<list> output in short format.

=head3 --long, -l

Display B<avail> and B<list> output in long format.

=head2 Module Sub-Commands

=head3 help [modulefile...]

Print the usage of each sub-command.
If an argument is given, print the Module specific help information for the
modulefile.

=head3 [load | add] modulefile [modulefile...]

Load modulefile into the shell environment.

=head3 [unload | rm] modulefile [modulefile...]

Remove modulefile from the shell environment.

=head3 [switch | swap] modulefile1 modulefile2

Switch loaded modulefile1 with modulefile2

=head3 [display | show] modulefile [modulefile...]

Display information about a modulefile.
The display sub-command will list the full path of the modulefile
and the environment changes the modulefile will make if loaded.  
(Note: It will not display any environment changes found within conditional
statements.)

=head3 list

List loaded modules.

=head3 avail [path...]

List all available modulefiles in the current \s-1MODULEPATH\s0.
All directories in the \s-1MODULEPATH\s0 are recursively searched for
files containing the modulefile magic cookie.  If an argument is
given, then each directory in the \s-1MODULEPATH\s0 is searched for 
modulefiles whose pathname match the argument.
Multiple versions of an application can be supported by creating a
subdirectory for the application containing modulefiles for each
version.

=head3 use [-a | --append] directory [directory...]

Prepend directory to the \s-1MODULEPATH\s0 environment variable.
The --append flag will append the directory to \s-1MODULEPATH\s0.

=head3 unuse directory [directory...]

Remove directory from the \s-1MODULEPATH\s0 environment variable.

=head3 update

Attempt to reload all loaded modulefiles.  The environment will be
reconfigured to match the saved .I ${\s-1HOME\s0}/.modulesbeginenv
and the modulefiles will be reloaded.  update will only change the
environment variables that the modulefiles set.

=head3 clear

Force the Modules Package to believe that no modules are currently loaded.

=head3 purge

Unload all loaded modulefiles.

=head3 whatis [modulefile [modulefile...]]

Display the modulefile information set up by the \fImodule-whatis\fP
commands inside the specified modulefiles. If no modulefiles are
specified all whatis information lines will be shown.

=head3 [apropos | keyword] string

Seeks thru the whatis informations of all modulefiles for the
specified string.  All module whatis informations matching the
string search will be displayed.

=head3 initadd modulefile [modulefile...]

Add modulefile to the shell's initialization file in the user's
home directory.  The startup files checked are .cshrc, .login,
and .csh_variables for the C Shell;
.profile for the Bourne and Korn Shells; .bashrc, .bash_env, 
and .bash_profile for the GNU Bourne Again Shell;
.zshrc, .zshenv, and .zlogin for zsh. The .modules file is checked
for all shells.

If a 'module load' line is found in any of these files, the
modulefile(s) are appended to any existing list of modulefiles.
The 'module load' line must be located in at least one of the
files listed above for any of the 'init' sub-commands to work properly.
If the 'module load' line is found in multiple shell initialization
files, all of the lines are changed.

=head3 initprepend modulefile [modulefile...]

Does the same as initadd but prepends the given modules to the
beginning of the list.

=head3 initrm modulefile [modulefile...]

Remove modulefile from the shell's initialization files.

=head3 initswitch modulefile1 modulefile2

Switch modulefile1 with modulefile2 in the shell's initialization files.

=head3 initlist

List all of the modulefiles loaded from the shell's initialization file.

=head3 initclear

Clear all of the modulefiles from the shell's initialization files.

=head1 Modulefiles

modulefiles are written in the Tool Command Language (tcl) and are
interpreted by modulecmd. modulefiles can use conditional statements.
Thus the effect a modulefile will have on the environment may change depending
upon the current state of the environment.

Environment variables are unset when unloading a modulefile.  Thus, it is
possible to load a modulefile and then unload it without having the
environment variables return to their prior state.

=head1 ENVIRONMENT

=head2 MODULESHOME

The location of the master Modules package file directory containing
module command initialization scripts, the executable program modulecmd,
and a directory containing a collection of master modulefiles.

=head2 MODULEPATH

The path that the module command searches when looking for modulefiles.
Typically, it is set to the master modulefiles directory,
${MODULESHOME}/modulefiles, by the initialization script.
MODULEPATH can be set using 'module use' or by the module
initialization script to search group or personal modulefile
directories before or after the master modulefile directory.

=head2 LOADEDMODULES

A colon separated list of all loaded modulefiles.

=head2 _LOADED_MODULEFILES_

A colon separated list of the full pathname for all loaded modulefiles.

=head2 _MODULESBEGINENV_

The filename of the file containing the initialization environment snapshot.

=head1 FILES

=head2 ${MODULESHOME}

The MODULESHOME directory.

=head2 ${MODULESHOME}/etc/rc

The system-wide modules rc file.  The location of this file can be changed
using the MODULERCFILE environment variable as described above.

=head2 ${HOME}/.modulerc

The user specific modules rc file.

=head2 ${MODULESHOME}/modulefiles

The directory for system-wide modulefiles.  The location of the directory
can be changed using the MODULEPATH environment variable as described above.

=head2 ${MODULESHOME}/bin/modulecmd

The modulefile interpreter that gets executed upon each invocation of module.

=head2 ${MODULESHOME}/init/shellname

The Modules package initialization file sourced into the user's environment.

=head1 SEE ALSO

B<modulefile>
